File: backend/models/patient.py
"""
Hasta veritabanı modeli

SQLAlchemy ORM modeli.
"""

from datetime import date, datetime
from typing import Optional, List
from sqlalchemy import Column, String, Date, DateTime, Text
from sqlalchemy.orm import relationship


# NOT: Base import'u gerçek uygulamada database.py'den yapılacak
# from database import Base

# Placeholder Base sınıfı
class BasePlaceholder:
    pass


Base = BasePlaceholder


class Patient:
    """
    Hasta veritabanı modeli

    Attributes:
        id: Benzersiz hasta ID'si (UUID)
        tc_kimlik_no: TC Kimlik Numarası (11 hane)
        ad_soyad: Hasta adı ve soyadı
        dogum_tarihi: Doğum tarihi
        cinsiyet: Cinsiyet (Erkek/Kadın)
        telefon: Telefon numarası (opsiyonel)
        email: E-posta adresi (opsiyonel)
        adres: Adres (opsiyonel)
        olusturma_tarihi: Kayıt oluşturulma tarihi
        guncelleme_tarihi: Son güncelleme tarihi
    """

    __tablename__ = "patients"

    # Birincil anahtar
    id: str  # Column(String(36), primary_key=True)

    # Zorunlu alanlar
    tc_kimlik_no: str  # Column(String(11), unique=True, nullable=False, index=True)
    ad_soyad: str  # Column(String(100), nullable=False, index=True)
    dogum_tarihi: date  # Column(Date, nullable=False)
    cinsiyet: str  # Column(String(10), nullable=False)

    # Opsiyonel alanlar
    telefon: Optional[str] = None  # Column(String(20), nullable=True)
    email: Optional[str] = None  # Column(String(100), nullable=True)
    adres: Optional[str] = None  # Column(Text, nullable=True)

    # Zaman damgaları
    olusturma_tarihi: datetime  # Column(DateTime, default=datetime.utcnow)
    guncelleme_tarihi: Optional[datetime] = None  # Column(DateTime, onupdate=datetime.utcnow)

    # İlişkiler
    # examinations = relationship("Examination", back_populates="patient")

    def __repr__(self) -> str:
        return f"<Patient(id={self.id}, tc={self.tc_kimlik_no}, ad_soyad={self.ad_soyad})>"

    @property
    def yas(self) -> int:
        """Hastanın yaşını hesapla"""
        today = date.today()
        age = today.year - self.dogum_tarihi.year
        if (today.month, today.day) < (self.dogum_tarihi.month, self.dogum_tarihi.day):
            age -= 1
        return age

    def to_dict(self) -> dict:
        """Modeli sözlüğe dönüştür"""
        return {
            "id": self.id,
            "tc_kimlik_no": self.tc_kimlik_no,
            "ad_soyad": self.ad_soyad,
            "dogum_tarihi": self.dogum_tarihi.isoformat() if self.dogum_tarihi else None,
            "cinsiyet": self.cinsiyet,
            "telefon": self.telefon,
            "email": self.email,
            "adres": self.adres,
            "yas": self.yas,
            "olusturma_tarihi": self.olusturma_tarihi.isoformat() if self.olusturma_tarihi else None,
            "guncelleme_tarihi": self.guncelleme_tarihi.isoformat() if self.guncelleme_tarihi else None,
        }


# Veritabanı şeması SQL (referans için)
"""
CREATE TABLE patients (
    id VARCHAR(36) PRIMARY KEY,
    tc_kimlik_no VARCHAR(11) UNIQUE NOT NULL,
    ad_soyad VARCHAR(100) NOT NULL,
    dogum_tarihi DATE NOT NULL,
    cinsiyet VARCHAR(10) NOT NULL,
    telefon VARCHAR(20),
    email VARCHAR(100),
    adres TEXT,
    olusturma_tarihi TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    guncelleme_tarihi TIMESTAMP
);

CREATE INDEX idx_patients_tc ON patients(tc_kimlik_no);
CREATE INDEX idx_patients_ad ON patients(ad_soyad);
"""

File: mobile/lib/models/patient.dart
/// Hasta veri modeli
class Patient {
  final String? id;
  final String tcKimlikNo;
  final String adSoyad;
  final DateTime dogumTarihi;
  final String cinsiyet;
  final String? telefon;
  final String? email;
  final String? adres;
  final DateTime? olusturmaTarihi;
  final DateTime? guncellemeTarihi;

  Patient({
    this.id,
    required this.tcKimlikNo,
    required this.adSoyad,
    required this.dogumTarihi,
    required this.cinsiyet,
    this.telefon,
    this.email,
    this.adres,
    this.olusturmaTarihi,
    this.guncellemeTarihi,
  });

  /// Yaş hesaplama
  int get yas {
    final now = DateTime.now();
    int age = now.year - dogumTarihi.year;
    if (now.month < dogumTarihi.month ||
        (now.month == dogumTarihi.month && now.day < dogumTarihi.day)) {
      age--;
    }
    return age;
  }

  /// JSON'dan Patient oluşturma
  factory Patient.fromJson(Map<String, dynamic> json) {
    return Patient(
      id: json['id'] as String?,
      tcKimlikNo: json['tc_kimlik_no'] as String,
      adSoyad: json['ad_soyad'] as String,
      dogumTarihi: DateTime.parse(json['dogum_tarihi'] as String),
      cinsiyet: json['cinsiyet'] as String,
      telefon: json['telefon'] as String?,
      email: json['email'] as String?,
      adres: json['adres'] as String?,
      olusturmaTarihi: json['olusturma_tarihi'] != null
          ? DateTime.parse(json['olusturma_tarihi'] as String)
          : null,
      guncellemeTarihi: json['guncelleme_tarihi'] != null
          ? DateTime.parse(json['guncelleme_tarihi'] as String)
          : null,
    );
  }

  /// Patient'ı JSON'a dönüştürme
  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'tc_kimlik_no': tcKimlikNo,
      'ad_soyad': adSoyad,
      'dogum_tarihi': dogumTarihi.toIso8601String().split('T')[0],
      'cinsiyet': cinsiyet,
      if (telefon != null) 'telefon': telefon,
      if (email != null) 'email': email,
      if (adres != null) 'adres': adres,
    };
  }

  /// Kopyalama metodu
  Patient copyWith({
    String? id,
    String? tcKimlikNo,
    String? adSoyad,
    DateTime? dogumTarihi,
    String? cinsiyet,
    String? telefon,
    String? email,
    String? adres,
  }) {
    return Patient(
      id: id ?? this.id,
      tcKimlikNo: tcKimlikNo ?? this.tcKimlikNo,
      adSoyad: adSoyad ?? this.adSoyad,
      dogumTarihi: dogumTarihi ?? this.dogumTarihi,
      cinsiyet: cinsiyet ?? this.cinsiyet,
      telefon: telefon ?? this.telefon,
      email: email ?? this.email,
      adres: adres ?? this.adres,
      olusturmaTarihi: olusturmaTarihi,
      guncellemeTarihi: guncellemeTarihi,
    );
  }

  @override
  String toString() {
    return 'Patient(id: $id, tcKimlikNo: $tcKimlikNo, adSoyad: $adSoyad, yas: $yas)';
  }
}
